src/react/components.js

```diff
@@ -273,15 +275,85 @@ export function applyGetDerivedStateFromProps(
+        } else if (a === null) {
+          invariant(b instanceof Value);
+          return AbstractValue.createFromConditionalOp(realm, condition, realm.intrinsics.false, b);
+        } else if (b === null) {
+          invariant(a instanceof Value);
+          return AbstractValue.createFromConditionalOp(realm, condition, a, realm.intrinsics.false);
+        } else {
+          invariant(a instanceof Value);
+          invariant(b instanceof Value);
+          return AbstractValue.createFromConditionalOp(realm, condition, a, b);
+        }
+      } else if (kind === "||" || kind === "&&") {
+        let a = deriveState(state.args[0]);
+        let b = deriveState(state.args[1]);
+        if (b === null) {
+          invariant(a instanceof Value);
+          return AbstractValue.createFromLogicalOp(realm, kind, a, realm.intrinsics.false);
+        } else {
+          invariant(a instanceof Value);
+          invariant(b instanceof Value);
+          return AbstractValue.createFromLogicalOp(realm, kind, a, b);
+        }
+      } else {
+        invariant(state.args !== undefined, "TODO: unknown abstract value passed to deriveState");
+        // as the value is completely abstract, we need to add a bunch of
+        // operations to be emitted to ensure we do the right thing at runtime
+        let a = AbstractValue.createFromBinaryOp(realm, "!==", state, realm.intrinsics.null);
+        let b = AbstractValue.createFromBinaryOp(realm, "!==", state, realm.intrinsics.undefined);
+        let c = AbstractValue.createFromLogicalOp(realm, "&&", a, b);
+        invariant(c instanceof AbstractValue);
+        let newState = new ObjectValue(realm, realm.intrinsics.ObjectPrototype);
+        // we cannot use the standard Object.assign as partial state
+        // is not simple. however, given getDerivedStateFromProps is
+        // meant to be pure, we can assume that there are no getters on
+        // the partial abstract state
+        AbstractValue.createTemporalFromBuildFunction(
+          realm,
+          FunctionValue,
+          [objectAssign, newState, prevState, state],
+          ([methodNode, ..._args]) => {
+            return t.callExpression(methodNode, ((_args: any): Array<any>));
+          }
+        );
+        newState.makeSimple();
+        newState.makePartial();
+        let conditional = AbstractValue.createFromLogicalOp(realm, "&&", c, newState);
+        return conditional;
+      }
+    } else if (state !== realm.intrinsics.null && state !== realm.intrinsics.undefined) {
+      let newState = new ObjectValue(realm, realm.intrinsics.ObjectPrototype);
+      objectAssignCall(realm.intrinsics.undefined, [newState, prevState, state]);
+      return newState;
+    } else {
+      return null;
+    }
+  };
+
+  let newState = deriveState(partialState);
+  if (newState !== null) {
+    if (newState instanceof AbstractValue) {
+      newState = AbstractValue.createFromLogicalOp(realm, "||", newState, prevState);
```

Comment:
Instead of adding this extra conditional, you could just make all the "else" branches where you return false, just return the prevState instead.

Reply:
I did that originally, I can go back to it.